### PR TITLE
Add outdated version warning and support for self-updating

### PR DESCRIPTION
**Pull request by axbarsan**
Closes https://github.com/giantswarm/roadmap/issues/533

This PR adds 3 things:
- Checking for updates after running any commands (except for `selfupdate` subcommands)
	- display a warning if there's a new version available
	- the warning is printed in `stderr`, so it's outputted to the terminal when using pipes, and is not forwarded to other commands
	- can be skipped by using the flag `--disable-version-check`
- Command for checking for updates (`kubectl gs selfupdate --dry-run`)
- Command for updating to the latest version (`kubectl gs selfupdate`)
	- works by fetching the current platform executable from the latest GitHub release, and replacing the executable in place

## Preview

<details>
<summary>Warning when using an outdated version</summary>

image

</details>

<details>
<summary>Checking for a new version - newer version available</summary>

image

</details>

<details>
<summary>Checking for a new version - already on latest</summary>

image

</details>

<details>
<summary>Updating - newer version available</summary>

image

</details>

<details>
<summary>Checking for a new version - already on latest</summary>

image

</details>

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -3,6 +3,7 @@ module github.com/giantswarm/kubectl-gs
 go 1.16
 
 require (
+	github.com/blang/semver v3.5.1+incompatible
```
Comment (axbarsan) [inline go.mod:6]: This is required to interface with `github.com/rhysd/go-github-selfupdate`. It was a transitive dependency before, and was linked anyway, so no big difference.

**Review by anvddriesch — APPROVED** [review]
LGTM

**Comment by marians** [comment]
The primary usecase here I think is "ensure that kubectl gs is up-to-date". I think that the command for that should ideally be shorter, so easier to memorize and to type.

This PR suggests

```nohighlight
kubectl gs selfupdate execute
```

I suggest instead

```nohighlight
kubectl gs selfupdate
```

If we want to offer a way to call the command for informational purposes only ("Show me whether my kubectl gs is up-to-date") we could do one of this:

```nohighlight
kubectl gs selfupdate --check-only
kubectl gs selfupdate --dry-run
```

WDYT?

**Comment by marians** [comment]
The output from other commands informing about a newer version available look fine to me.

File: CHANGELOG.md [hunk 2]
```diff
@@ -7,6 +7,12 @@ and this project's packages adheres to [Semantic Versioning](http://semver.org/s
 
 ## [Unreleased]
 
+### Added
+
+- Print warning after running any command if there is a newer version available.
+- Implement command for checking for newer versions.
+- Implement command for self-updating.
```
Comment (marians) [inline CHANGELOG.md:14]: Would be nice to actually mention the commands once that is finally decided.

**Comment by marians** [comment]
@axbarsan Do you have any response to my feedback?

**Comment by ubergesundheit** [comment]
For people installing kubectl-gs through krew, will it be possible to disable the upgrade check?

**Comment by axbarsan** [comment]
> For people installing kubectl-gs through krew, will it be possible to disable the upgrade check?

It is possible to use the `--disable-version-check` flag to disable it

**Comment by marians** [comment]
I'm curious whether users will ask for something more permanent, like an environment variable. But I'd be willing to wait for such a request after this is released.

**Comment by ubergesundheit** [comment]
> It is possible to use the --disable-version-check flag to disable it

> I'm curious whether users will ask for something more permanent, like an environment variable. But I'd be willing to wait for such a request after this is released.

Yeah, I was looking for something more persistent

**Comment by axbarsan** [comment]
> The primary usecase here I think is "ensure that kubectl gs is up-to-date". I think that the command for that should ideally be shorter, so easier to memorize and to type.
> 
> This PR suggests
> 
> ```
> kubectl gs selfupdate execute
> ```
> 
> I suggest instead
> 
> ```
> kubectl gs selfupdate
> ```
> 
> If we want to offer a way to call the command for informational purposes only ("Show me whether my kubectl gs is up-to-date") we could do one of this:
> 
> ```
> kubectl gs selfupdate --check-only
> kubectl gs selfupdate --dry-run
> ```
> 
> WDYT?

Fine with me

**Comment by axbarsan** [comment]
The command was updated to use `kubectl gs selfupdate` for updating and `kubectl gs selfupdate --dry-run` for checking for updates

**Review by marians — APPROVED** [review]
Nice, thank you!

**Pull request merged**